test/e2e/garbage_collector.go

```diff
@@ -441,4 +472,239 @@ var _ = framework.KubeDescribe("Garbage collector", func() {
+		if err != nil {
+			framework.Failf("Failed to create replication controller: %v", err)
+		}
+		rc2Name := "simpletest-rc-to-stay"
+		rc2 := newOwnerRC(f, rc2Name, 0)
+		rc2.Spec.Selector = nil
+		rc2.Spec.Template.ObjectMeta.Labels = map[string]string{"another.key": "another.value"}
+		By("create the rc2")
+		rc2, err = rcClient.Create(rc2)
+		if err != nil {
+			framework.Failf("Failed to create replication controller: %v", err)
+		}
+		// wait for rc1 to be stable
+		if err := wait.Poll(5*time.Second, 30*time.Second, func() (bool, error) {
+			rc1, err := rcClient.Get(rc1.Name, metav1.GetOptions{})
+			if err != nil {
+				return false, fmt.Errorf("Failed to get rc: %v", err)
+			}
+			if rc1.Status.Replicas == *rc1.Spec.Replicas {
+				return true, nil
+			} else {
+				return false, nil
+			}
+		}); err != nil {
+			framework.Failf("failed to wait for the rc.Status.Replicas to reach rc.Spec.Replicas: %v", err)
+		}
+		By(fmt.Sprintf("set half of pods created by rc %s to have rc %s as owner as well", rc1Name, rc2Name))
+		pods, err := podClient.List(metav1.ListOptions{})
+		Expect(err).NotTo(HaveOccurred())
+		patch := fmt.Sprintf(`{"metadata":{"ownerReferences":[{"apiVersion":"v1","kind":"ReplicationController","name":"%s","uid":"%s"}]}}`, rc2.ObjectMeta.Name, rc2.ObjectMeta.UID)
+		for i := 0; i < halfReplicas; i++ {
+			pod := pods.Items[i]
+			_, err := podClient.Patch(pod.Name, types.StrategicMergePatchType, []byte(patch))
+			Expect(err).NotTo(HaveOccurred())
+		}
+
+		By(fmt.Sprintf("delete the rc %s", rc1Name))
+		deleteOptions := getForegroundOptions()
+		deleteOptions.Preconditions = metav1.NewUIDPreconditions(string(rc1.UID))
+		if err := rcClient.Delete(rc1.ObjectMeta.Name, deleteOptions); err != nil {
+			framework.Failf("failed to delete the rc: %v", err)
+		}
+		By("wait for the rc to be deleted")
+		// default client QPS is 20, deleting each pod requires 2 requests, so 30s should be enough
+		if err := wait.Poll(5*time.Second, 30*time.Second, func() (bool, error) {
+			_, err := rcClient.Get(rc1.Name, metav1.GetOptions{})
+			if err == nil {
+				pods, _ := podClient.List(metav1.ListOptions{})
+				framework.Logf("%d pods remaining", len(pods.Items))
+				count := 0
+				for _, pod := range pods.Items {
+					if pod.ObjectMeta.DeletionTimestamp == nil {
+						count++
+					}
+				}
+				framework.Logf("%d pods has nil DeletionTimestamp", count)
+				framework.Logf("")
+				return false, nil
+			} else {
+				if errors.IsNotFound(err) {
+					return true, nil
+				} else {
+					return false, err
+				}
+			}
+		}); err != nil {
+			pods, err2 := podClient.List(metav1.ListOptions{})
+			if err2 != nil {
+				framework.Failf("%v", err2)
+			}
+			framework.Logf("%d remaining pods are:", len(pods.Items))
+			framework.Logf("ObjectMeta of remaining pods are:")
+			for _, pod := range pods.Items {
+				framework.Logf("%#v", pod.ObjectMeta)
+			}
+			framework.Failf("failed to delete rc %s, err: %v", rc1Name, err)
```

Comment:
rc1 should get deleted.

rc1 initially has 100 dependents, 50 of which also have rc2 as an owner. rc2 is not deleted. When a foreground deletion of rc1 is issued, rc1 should be deleted after the 50 pods that solely depend on rc1 are deleted. The remaining pods should only have rc2 as the owner.

This behavior is described in the [proposal](https://github.com/kubernetes/community/blob/master/contributors/design-proposals/synchronous-garbage-collection.md), starts with "if an object has multiple owners...".